### PR TITLE
[FLINK-8472] [test] Extend all migration tests for Flink 1.4

### PR DESCRIPTION
**Pull request by tzulitai**
## What is the purpose of the change

Extend all migration tests, to include verifying restore from Flink 1.4 savepoints.

This includes extending:
- `WindowOperatorMigrationTest`
- `CEPMigrationTest`
- `StatefulJobSavepointMigrationTestITCase` (Scala API migration)
- `StatefulJobSavepointMigrationTestITCase` (Java API migration)
- `FlinkKinesisConsumerMigrationTest`
- `FlinkKafkaConsumerBaseMigrationTest`
- `ContinuousFileProcessingMigrationTest`
- `BucketingSinkMigrationTest`

This PR should also be forward-ported to the `master` branch to cover Flink 1.5.

## Brief change log

All commits except 1ce3e6c are simply adding `MigrationVersion.v1_4` to the test parameters and adding the corresponding test savepoint files.

1ce3e6c is a refactor of `StatefulJobSavepointMigrationFrom12ITCase` and `StatefulJobSavepointMigrationFrom13ITCase` to a single `StatefulJobSavepointMigrationITCase` class.

## Verifying this change

This is a test refactor / extension, so all existing tests should cover the changes.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by tzulitai** [comment]
It seems like we'll also need to extend these migration tests for 1.4:

- All subclasses of `AbstractKeyedOperatorRestoreTestBase`
- All subclasses of `AbstractNonKeyedOperatorRestoreTestBase`

**Comment by aljoscha** [comment]
The changes look straightforward and good! 👍 

Plus, they help me sleep better... 😅

**Comment by tzulitai** [comment]
@aljoscha @zentol 
could you have another quick look at commit 8882fb7? That commit extends `KeyedComplexChainTest`, `ChainBreakTest`, `ChainLengthIncreaseTest`, ... etc. also for 1.4.

Once that is good, I'll merge this to `master` and `release-1.4`.

File: flink-tests/src/test/java/org/apache/flink/test/state/operator/restore/keyed/KeyedJob.java [hunk 1]
```diff
@@ -100,9 +100,7 @@ public static void main(String[] args) throws Exception {
 			.map(new StatefulStringStoringMap(mode, "first"))
 			.setParallelism(4);
 
-		if (mode == ExecutionMode.MIGRATE || mode == ExecutionMode.RESTORE) {
```
Comment (tzulitai) [inline flink-tests/src/test/java/org/apache/flink/test/state/operator/restore/keyed/KeyedJob.java]: @zentol can you confirm whether these changes make sense?

From the discussions I see in https://github.com/apache/flink/pull/3844, I assume this was a leftover to write a 1.2 savepoint (when uids couldn't be added for each chained operator separately). So it should be ok to remove this?

**Comment by aljoscha** [comment]
Looks good but let's wait what @zentol has to say.

File: flink-tests/src/test/java/org/apache/flink/test/state/operator/restore/keyed/KeyedJob.java (hunk 1 above)
Comment (zentol) [reply flink-tests/src/test/java/org/apache/flink/test/state/operator/restore/keyed/KeyedJob.java]: yeah it should be alright to remove that, but let's chain the uid call to the operator creation as we do for the others for style points.

**Comment by tzulitai** [comment]
Thanks for the reviews, I'll address Chesnay's comment and the merge this then (to `master` and `release-1.4`).

**Pull request closed without merging**